### PR TITLE
Rewrite CFNumberGetValue tests for OSX

### LINKED ISSUE #758: NSNumber.Derived_CFNumberGetValue fails on OSX
[ RUN      ] NSNumber.Derived_CFNumberGetValue
Foundation/NSNumberTests.mm:204: Failure
Value of: CFNumberGetValue(static_cast<CFNumberRef>(num), kCFNumberLongType, &tooSmallResult)
  Actual: true
Expected: false
[  FAILED  ] NSNumber.Derived_CFNumberGetValue (0 ms)

### PR DESCRIPTION
**Pull request by ms-jihua**
- OSX functionality for CFNumberGetValue seems to not follow its documentation,
   not returning false when the return value is out of range,
   nor returning false when the return value is lossy, and the instance is a subclass.

Expanded the test to multiple tests, disabled some on OSX - It seems better to adhere to the documentation in at least this case.

Fixes #758

### CODE REVIEW

**Comment by ms-jihua** [comment]
@DHowett-MSFT is added to the review. #Closed

**Comment by ms-jihua** [comment]
@mnithish is added to the review. #Closed

**Comment by ms-jihua** [comment]
@MSFTFox is added to the review. #Closed

**Comment by DHowett-MSFT** [comment]
:shipit:

File: tests/unittests/Foundation/NSNumberTests.mm [hunk 1]
```diff
@@ -191,15 +191,81 @@ - (const char*)objCType {
...
+    NSNumber* num = [TestNSNumberFloatSubclass numberWithFloat:expected];
+
+    float result;
+    ASSERT_TRUE(CFNumberGetValue(static_cast<CFNumberRef>(num), kCFNumberFloatType, &result));
+    ASSERT_EQ(expected, result);
+
+    // Loss of floating-point precision should return false
+    long integralResult;
+    ASSERT_FALSE(CFNumberGetValue(static_cast<CFNumberRef>(num), kCFNumberLongType, &integralResult));
 }
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSNumberTests.mm]: Is there room to determine whether a double-precision value truncated into a float-precision location returns false?

Comment (ms-jihua) [inline tests/unittests/Foundation/NSNumberTests.mm]: Discussed offline, we don't think this is a fruitful line of investigation at this time.

---
In reply to: [73772452](https://github.com/Microsoft/WinObjC/pull/829#discussion_r73772452) [](ancestors = 73772452)

**Comment by ms-jihua** [comment]
I seem to have gotten an unrelated test failure in CI build:

FAILING TESTS:
   Mach.MachTaskInfoThreadTimes  ( D:\Build\Agent3\_work\7\s\build\Win32\Debug\WOCStdLib.UnitTests\WOCStdLib.UnitTests.exe --gtest_filter=Mach.MachTaskInfoThreadTimes )

Re-queueing, but is the above a known issue?

**Comment by ms-jihua** [comment]
@rajsesh-msft is added to the review. #Closed

**Pull request merged**